chasm/lib/workflow/workflow.go

```diff
@@ -37,3 +45,84 @@ func (w *Workflow) LifecycleState(
 	// Otherwise, tasks for callback component can't be executed after workflow is closed.
 	return chasm.LifecycleStateRunning
 }
+
+// ProcessCloseCallbacks triggers "WorkflowClosed" callbacks using the CHASM implementation.
+// It iterates through all callbacks and schedules WorkflowClosed ones that are in STANDBY state.
+func (w *Workflow) ProcessCloseCallbacks(ctx chasm.MutableContext) error {
+	// Iterate through all callbacks and schedule WorkflowClosed ones
+	for _, field := range w.Callbacks {
+		cb := field.Get(ctx)
+		// Only process callbacks in STANDBY state (not already triggered)
+		if cb.Status != callbackspb.CALLBACK_STATUS_STANDBY {
+			continue
+		}
+		// Trigger the callback by transitioning to SCHEDULED state
+		if err := callback.TransitionScheduled.Apply(ctx, cb, callback.EventScheduled{}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AddCompletionCallbacks creates completion callbacks using the CHASM implementation.
+// maxCallbacksPerWorkflow is the configured maximum number of callbacks allowed per workflow.
+func (w *Workflow) AddCompletionCallbacks(
+	ctx chasm.MutableContext,
+	eventTime *timestamppb.Timestamp,
+	requestID string,
+	completionCallbacks []*commonpb.Callback,
+	maxCallbacksPerWorkflow int,
+) error {
+	// Check CHASM max callbacks limit
+	currentCallbackCount := len(w.Callbacks)
+	if len(completionCallbacks)+currentCallbackCount > maxCallbacksPerWorkflow {
+		return serviceerror.NewFailedPreconditionf(
+			"cannot attach more than %d callbacks to a workflow (%d callbacks already attached)",
+			maxCallbacksPerWorkflow,
+			currentCallbackCount,
+		)
+	}
+
+	// Initialize map if needed
+	if w.Callbacks == nil {
+		w.Callbacks = make(chasm.Map[string, *callback.Callback], len(completionCallbacks))
+	}
+
+	// Add each callback
+	for idx, cb := range completionCallbacks {
+		chasmCB := &callbackspb.Callback{
+			Links: cb.GetLinks(),
+		}
+		switch variant := cb.Variant.(type) {
+		case *commonpb.Callback_Nexus_:
+			chasmCB.Variant = &callbackspb.Callback_Nexus_{
+				Nexus: &callbackspb.Callback_Nexus{
+					Url:    variant.Nexus.GetUrl(),
+					Header: variant.Nexus.GetHeader(),
+				},
+			}
+		default:
+			return fmt.Errorf("unsupported callback variant: %T", variant)
+		}
+
+		id := fmt.Sprintf("cb-%s-%d", requestID, idx)
```

Comment:
You can save a couple of bytes on the prefix.

```suggestion
		id := fmt.Sprintf("%s-%d", requestID, idx)
```